firestore/bulkwriter.go

```diff
@@ -0,0 +1,319 @@
+package firestore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	vkit "cloud.google.com/go/firestore/apiv1"
+	"google.golang.org/api/support/bundler"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+)
+
+const (
+	// maxBatchSize is the max number of writes to send in a request
+	maxBatchSize = 20
+	// retryMaxBatchSize is the max number of writes to send in a retry request
+	retryMaxBatchSize = 10
+	// maxRetryAttempts is the max number of times to retry a write
+	maxRetryAttempts = 10
+	// defaultStartingMaximumOpsPerSecond is the starting max number of requests to the service per second
+	defaultStartingMaximumOpsPerSecond = 500
+)
+
+// BulkWriterJob provides read-only access to the results of a BulkWriter write attempt.
+type BulkWriterJob struct {
+	err      chan error           // send error responses to this channel
+	result   chan *pb.WriteResult // send the write results to this channel
+	write    *pb.Write            // the writes to apply to the database
+	attempts int                  // number of times this write has been attempted
+	wr       *WriteResult         // (cached) result from the operation
+	e        error                // (cached) any errors that occurred
+	ctx      context.Context      // context for canceling/timing out results
+}
+
+// Results gets the results of the BulkWriter write attempt.
+// This method blocks if the results for this BulkWriterJob haven't been
+// received.
+func (j *BulkWriterJob) Results() (*WriteResult, error) {
+	if j.wr == nil && j.e == nil {
+		j.wr, j.e = j.processResults() // cache the results for additional calls
+	}
+	return j.wr, j.e
+}
+
+// processResults checks for errors returned from send() and packages up the
+// results as WriteResult objects
+func (j *BulkWriterJob) processResults() (*WriteResult, error) {
+	select {
+	case <-j.ctx.Done():
+		return nil, fmt.Errorf("bulkwriter: early write cancellation")
+	case wpb := <-j.result:
+		return writeResultFromProto(wpb)
+	case err := <-j.err:
+		return nil, err
+	}
+}
+
+// setError ensures that an error is returned on the error channel of BulkWriterJob.
+func (j *BulkWriterJob) setError(e error) {
+	j.err <- e
+}
+
+// setSuccess ensures that a WriteResult is returned to the result channel of BulkWriterJob.
+func (j *BulkWriterJob) setResult(r *pb.WriteResult) {
```

Comment:
This is only used in one spot, I would suggest just inlining.

Reply:
As written, I agree. If the intention is to also close the error channel (or the result channel above), then it might be work keeping these around. In any case, don't we need to close the other channel when we set a result or an error?

Reply:
I think I added these based upon a previous round of feedback 😄 . You're right, I need to close the other channels when I return.